runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -418,6 +418,19 @@ public void stop() {
         }
     }
 
+    /**
+     * Reestablish the netty connection.
+     */
+    @Override
+    public void reconnect() {
+        // Close the channel and auto-trigger the reconnection callback.
+        log.info("NettyClientRouter reconnecting. Closing the existing channel.");
+        Channel channelCopy = channel;
+        if (channelCopy != null && channelCopy.isOpen()) {
+            channelCopy.close();
+        }
```

Comment:
It looks like there's a potential of leaking Netty native resources here as this interleaves with `stop()`.  Suppose that `reconnect()` is invoked for some client router. The channel is closed and the reconnection callback is invoked. As this occurs but before the new channel reference is updated in `NettyClientRouter::userEventTriggered`, `stop()` is invoked from the CorfuRuntime thread executing `fetchLayout` via `pruneRemovedRouters`. When this thread executes `stop()`, the channel is already closed, but then the reference is once again updated from the thread executing `userEventTriggered`.

I'm thinking of some scenario where a node is removed from the cluster and hence from the layout, but is still temporarily reachable via the network.

Reply:
I think even without the new `reconnect` API the issue you described already exists today. The root of the problem is that there is no synchronization between the `NettyClientRouter ::userEventTriggered ` and `NettyClientRouter ::stop`. Although `addReconnectionOnCloseFuture` and `connectAsync` checks the volatile flag `shutdown`, consider a case where `stop()` is called after `connectAsync` reads `shutdown` and `stop()` closes the channel before `userEventTriggered ` re-assigns the channel, it will be a connection leak. I think what you described is true but the impact of the leak is relatively small.